sql-plugin/src/main/scala/com/nvidia/spark/rapids/StrategyRules.scala

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids
+
+import com.nvidia.spark.rapids.delta.DeltaProvider
+
+import org.apache.spark.sql.Strategy
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.execution.SparkPlan
+
+/**
+ * Provides a Strategy that can implement rules for translating
+ * custom logical plan nodes to physical plan nodes.
+ * @note This is instantiated via reflection from ShimLoader.
+ */
+class StrategyRules extends Strategy {
+
+  private lazy val strategies: Seq[Strategy] = {
+    // Currently we only have custom plan nodes that originate from
+    // DeltaLake, but if we add other custom plan nodes later
+    // their strategies can be appended here.
+    DeltaProvider().getStrategyRules
+  }
+
+  override def apply(plan: LogicalPlan): Seq[SparkPlan] = {
+    val rapidsConf = new RapidsConf(plan.conf)
+    if (rapidsConf.isSqlEnabled && rapidsConf.isSqlExecuteOnGPU) {
+      // Using view since the strategies are first fit.
+      strategies.view.map(_(plan)).find(_.nonEmpty).getOrElse(Nil)
```

Comment:
So we apply the first rule to this that node not return Nil, and go with that, and if we cannot find one we return Nil. This feels a little odd. Is this the same thing that Spark does? If so then why is it a Sequence? If not, then why are we different?

Reply:
This is close to what Spark is doing.  See the `GenericStrategy` interface in Spark and Catalyst's `QueryPlanner` has a sequence of strategies that are called in turn.  We have to "play nice" with that, so we have to either match on the plan or return Nil.  I just delegated that to an internal list of strategies since it was complicated to register them all directly with Spark due to classloader issues trying to register them directly with Spark so early in the startup.

The "early out" is something I added since unlike Spark I didn't expect we would have strategies that built on top of the work of our other strategies.  Regarding keeping it a sequence like Spark, I didn't make it a hash map since I wanted to allow for more complicated pattern matching than just a single logical node.  We're not just handing a single node to the strategy, we're handing it the sub-plan starting at that node. However if we don't think we'll ever need anything than one-to-one I could make it a hashmap lookup similar to that used by GpuOverrides.